### ISSUE TITLE
Using WITH DISTINCT combined with OPTIONAL MATCH leads to exception

### ISSUE THREAD

**Issue by AlexiosP** [neo4j/neo4j#12774]
While trying to optimize & troubleshoot a more complex query, 
I found the issue that `distinct` is throwing an error when used with `optional match`.

- Neo4j version: 4.2.3 Enterprise 
- Operating system: macOS Mojave 10.14.6
- API/Driver: Cypher
### Steps to reproduce
Run cypher query 
```
OPTIONAL MATCH ()-[rel]-()
WITH DISTINCT rel
RETURN rel
```
### Expected behavior
`[]` or `[null]` on an empty database (in a database with data, a distinct list of the relationships possibly with `null` included in the list)
### Actual behavior
On an empty database it returns the exception
```
Expected 
RegularSinglePlannerQuery(QueryGraph {Optional Matches: : ['QueryGraph {Rels: ['(  NODE16)--[rel]--(  NODE25)']}']},InterestingOrder(RequiredOrderCandidate(List()),List()),AggregatingQueryProjection(Map(),Map(__sr__ -> FunctionInvocation(Namespace(List()),FunctionName(collect),true,Vector(Variable(rel)))),QueryPagination(None,None),Selections(Set())),Some(RegularSinglePlannerQuery(QueryGraph {Arguments: ['__sr__']},InterestingOrder(RequiredOrderCandidate(List()),List()),RegularQueryProjection(Map(__sr__ -> Variable(__sr__)),QueryPagination(None,None),Selections(Set())),None,None)),None) 


Instead, got: 
RegularSinglePlannerQuery(QueryGraph {Optional Matches: : ['QueryGraph {}']},InterestingOrder(RequiredOrderCandidate(List()),List()),AggregatingQueryProjection(Map(),Map(__sr__ -> FunctionInvocation(Namespace(List()),FunctionName(collect),true,Vector(Variable(rel)))),QueryPagination(None,None),Selections(Set())),Some(RegularSinglePlannerQuery(QueryGraph {Arguments: ['__sr__']},InterestingOrder(RequiredOrderCandidate(List()),List()),RegularQueryProjection(Map(__sr__ -> Variable(__sr__)),QueryPagination(None,None),Selections(Set())),None,None)),None)
Plan: ProduceResult(List(__sr__)) {
  LHS -> Apply() {
    LHS -> Aggregation(Map(), Map(__sr__ -> FunctionInvocation(Namespace(List()),FunctionName(collect),true,Vector(Variable(rel))))) {
      LHS -> Apply() {
        LHS -> Argument(Set()) {}
        RHS -> Optional(Set()) {
          LHS -> Argument(Set()) {}
        }
      }
    }
    RHS -> Argument(Set(__sr__)) {}
  }
} 


Differences:
 - QueryGraph
    A: QueryGraph {Optional Matches: : ['QueryGraph {Rels: ['(  NODE16)--[rel]--(  NODE25)']}']}
    B: QueryGraph {Optional Matches: : ['QueryGraph {}']}
```

**Comment by mnd999** [comment]
Thanks for the report, that appears to be a bug. We'll investigate and get back to you.

**Comment by AlexiosP** [comment]
Thanks! Just a correction, first I was trying to report the below
```
OPTIONAL MATCH ()-[rel]-()
WITH collect(DISTINCT rel) as rels
RETURN rels
```

so the expected behavior I have written in the original post was for the above query. 
Afterwards I realized I could simplify the query that causes the exception further.
Without the `collect` the expected behavior would be `null` in case of an empty database, or (in a database with data) multiple rows instead of a list.

**Comment by sherfert** [comment]
@AlexiosP 

This has been fixed in [4.2.10](https://github.com/neo4j/neo4j/wiki/Neo4j-4.2-changelog#4210) and [4.3.3](https://github.com/neo4j/neo4j/wiki/Neo4j-4.3-changelog#433).
